web/interceptor/access_control.go

```diff
@@ -0,0 +1,150 @@
+package interceptor
+
+import (
+	"context"
+	"strings"
+
+	"github.com/quickfeed/quickfeed/qf"
+	"github.com/quickfeed/quickfeed/web/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+type (
+	role      int
+	roles     []role
+	requestID interface {
+		FetchID(string) uint64
```

Comment:
Please refactor this as follows:
```go
	requestID interface {
		IDFor(role) uint64
	}
```
See below for additional suggestions.

Reply:
Wouldn't I need to move enum with roles into the `qf` package and make it exported for this to work? That would be problematic because `User`, `Group`, and other names are already defined there.

Reply:
You are right. However, I found a trick, that I think will work for us...

1. Move / copy these to `qf/roles.go` (I've removed the comments from here; feel free to keep them):

```go
type (
	Role  int
	Roles []Role
)

const (
	none Role = iota
	user
	group
	student
	teacher
	courseAdmin
	admin
)
```

2. Change the interface like this in `access_control.go`:

```go
type requestID interface {
	IDFor(qf.Role) uint64
}
```

3. Update this in `access_control.go`:

```go
const (
	none qf.Role = iota
(the rest of the consts doesn't need to change)
...
var access = map[string]qf.Roles{
...
(and possibly a few other places)
```

This should do the trick. The only thing to keep in mind is that the consts are now in two files that must be synced. It is probably a good idea to make a cross-reference comment pointing to both files.

Reply:
I've given this some more thought, given my discovery late last night that the `IDFor(string)` approach isn't really expecting a `role` as input (there is no `course` role...) Thus, it seems inappropriate to use my proposed approach with `IDFor(qf.Role)`.

What had me confused were these implementations that seem to indicate that these were the same roles as those in `access_control.go`. But they are not.
```go
func (r *GroupRequest) IDFor(role string) uint64 { }

func (r *SubmissionRequest) IDFor(role string) uint64 { }
```

I suggest to leave it as `IDFor(string)` for now. We should probably reconsider this in the context of [this suggestion to merge some proto messages](https://github.com/quickfeed/quickfeed/issues/701#issuecomment-1202409979).